### ISSUE TITLE
pass4symkey uses default value

### ISSUE THREAD

**Issue by mikedickey** [splunk/splunk-ansible#316]
Ansible plays do not appear to be using the value of splunk.shc.secret from defaults.yml.

```
splunk:
    shc:
        secret: "K5B9OWb3XzDCwDdPVuZMMmAE"
```

```
$ head -3 /opt/splunk/etc/system/local/server.conf
[general]
serverName = splunk-skynet-mini-cluster-master-0
pass4SymmKey = $7$qtOAaLwaELHHHcL9ihnKOQ0eMJvkxrCiMtIg6HN0cXk3baNj7LRJAQ==
```

```
$ /opt/splunk/bin/splunk show-decrypted --value '$7$qtOAaLwaELHHHcL9ihnKOQ0eMJvkxrCiMtIg6HN0cXk3baNj7LRJAQ=='
changeme
```

**Comment by mikedickey** [comment]
This manifests in an _internal index error that looks something like this:

```
12-11-2019 04:18:33.574 +0000 WARN  SHCConfig - Default pass4symkey is being used. Please change to a random one.
```

After a little more digging, this has something to do with the role..  In particular, the error comes from the cluster master, license master, and all indexers.

On the search heads, the custom key is getting pulled from defaults.yml and used to seed the secret within the [shclustering] stanza of server.conf. So no errors come from those.

**Comment by mikedickey** [comment]
FYI i tried adding a splunk.secret value to my defaults and it causes everything to fail:

```
TASK [splunk_common : Set the Splunk secret] ***********************************
fatal: [localhost]: FAILED! => {}

MSG:

Failed to get information on remote file (/opt/splunk/etc/auth/splunk.secret): sudo: a password is required
```

**Comment by jrigassio-splunk** [comment]
@nwang92  helped me figure out what might be going on here. Based on this https://answers.splunk.com/answers/639815/what-is-the-splunksecret-file-and-is-it-possible-t.html it looks like that secret is used to encrypt the `pass4Symkey` values in the `[general]`, `[sshclustering]`, and `[clustering]` stanzas of `/opt/splunk/etc/system/local/server.conf`. 

It seems like we can include a new option in default.yml to overwrite the `pass4Symkey` field to address this, rather than update `pass4Symkey` with the value of `splunk.secret`. We could even add another for `sslPassword` (which is currently the default value `password`)

**Comment by mikedickey** [comment]
Not sure I'm completely following this, but generally I think fewer (versus more) options is preferable, whenever possible. Just want to be clear there appear to be two problems here:

* The value of splunk.shc.secret in defaults.yml is not being used to set the pass4Symkey value, as according to documentation it should be.

* Setting a custom splunk.secret value in defaults.yml appears to be broken

The original error message that triggered this investigation was "Default pass4symkey is being used" and this appears to be caused by the first problem.

**Comment by nwang92** [comment]
@jrigassio-splunk and I have found that `splunk.shc.secret` does indeed set the `pass4SymmKey` correctly, but only for the `pass4SymmKey` of the `shclustering` stanza in server.conf. So given a default.yml which consists of:
```
splunk:
  shc:
    secret: helloworld
  ...
```
A proper SHC will have the following in `/opt/splunk/etc/local/server.conf`:
```
[shclustering]
pass4SymmKey = <ecrypted value of "helloworld">
...
```

The problem here is more of a semantics issue, where `splunk.secret` does not behave in a similar manner as `splunk.shc.secret`. While the value of `splunk.shc.secret` gets directly mapped to the value for `pass4SymmKey`, the value of `splunk.secret` is used to create the `/opt/splunk/etc/auth/splunk.secret` file. The passphrase in `/opt/splunk/etc/auth/splunk.secret` is **then** used to perform all the encryption for all the other sensitive information (including pass4SymmKey) so conf files on Splunk's filesystem are secure.

So given a default.yml that contains:
```
splunk:
  secret: thisismysecret
  shc:
    secret: helloworld
  ...
```
What ends up happening is the plaintext "helloworld" gets encrypted _using_ the key "thisismysecret" and turns into the value you see when you inspect the contents of `[shclustering]` in server.conf.

**Comment by nwang92** [comment]
To get rid of the warning you observed, my guess is we'd need to change the `pass4SymmKey` in the `[general]` stanza of server.conf before Splunk starts. Currently we don't expose any option for this - we could reuse the existing `splunk.secret` variable but this seems a little misleading.

Additionally if the default Splunk `pass4SymmKey` of `[general]` is a potential security/usability problem, server.conf also ships with a default `sslPassword` in `[sslConfig]` of server.conf which might also warrant being changed? If you run the `splunk show-decrypted` command on this, you'll see it has a default value of just "password".

**Comment by mikedickey** [comment]
Thanks, Nelson, I think it's starting to make more sense to me now! This is all a bit muddied I think by the mismatch between parameter names `pass4SymKey` and `secret` in defaults.yml.

How about the following proposal:

* We add a new `splunk.pass4SymmKey` parameter that sets the value in `[general]`
* We rename the parameters `splunk.*.secret` to `splunk.*.pass4SymmKey`. We should allow `splunk.*.secret` to continue to be used for backwards-compatibility
* We update the docs for defaults.yml to be more clear on all of these

For example, `splunk.shc.secret` would become `splunk.shc.pass4SymmKey` and it would set the `pass4SymmKey` parameter in `[shclustering]`. 

There is still the matter of it failing when `splunk.secret` is defined. Looking back, I realized it's probably because I'm running the container as a regular user and the play requires sudo. Any idea why it would have that requirement? If not, can we please remove it?

**Comment by nwang92** [comment]
That sounds good to me, I'll look into the permissions issues on `splunk.secret` and see if I can remove the requirement on sudo.

**Comment by nwang92** [comment]
Permissions should be fixed here: https://github.com/splunk/splunk-ansible/pull/343 - @jrigassio-splunk will take care of the other additions

**Comment by nwang92** [comment]
So the sslPassword in `[sslConfig]` is a bit tricky - I think the expectation here for this setting aimed primarily at users who are bringing their own self-signed or outside certificates to encrypt Splunk-to-Splunk communication. However, if you simply change the provided sslPassword "password" that seems to be invalidating the default certs shipped in ${SPLUNK_HOME}/etc/auth/ and preventing Splunk from starting.

Anyways, I'm going to close this issue now since we opened up https://github.com/splunk/splunk-ansible/issues/357 to address any future customers who want to bring their own certificates.

**Comment by mikedickey** [comment]
I believe this doc file also needs to be updated: https://github.com/splunk/splunk-ansible/blob/develop/docs/advanced/default.yml.spec.md

**Comment by mikedickey** [comment]
Reopening since backwards compat appears to be not working quite right

**Comment by jrigassio-splunk** [comment]
Closing the issue now with backwards compat fix and doc added in #368.

**Comment by henricook** [comment]
I needed this today and for some reason couldn't find it in any of the docs, has it been deprecated or has there perhaps been an update issue?

I've guesstimated that my default.yaml should look like this, is it right?

```
---
splunk:
  http_port: 8001
  hec:
    ssl: false
  pass4SymmKey: PASSGOESHERE
```
